### PR TITLE
Add economic viability check in the driver

### LINKED ISSUE #2613: feat: Add economic viability check in the driver
This was requested from a solver team and asks for the following feature that could be enabled with a flag:

When the driver simulates the solution, it checks whether `estimated_cost > network_fees`, and if so, replaces the proposed solution with the empty one.

### PR DESCRIPTION
**Pull request by Mateo-mro**
# Description
When the driver simulates the solution, it checks whether `simulated_gas` > `solver_reported_gas`, and if so, replaces the proposed solution with the empty one.

# Changes
Return a new type of `Revert` error if the simulated gas hits the gas limit.

## How to test
1. regression tests

## Related Issues

Fixes #2613

### CODE REVIEW

File: crates/driver/src/infra/simulator/mod.rs
Comment (squadgazzz) [inline crates/driver/src/infra/simulator/mod.rs:155]:
```suggestion
    #[error("the simulated gas {0:?} exceeded the gas limit {1:?}")]
```

Comment (Mateo-mro) [reply crates/driver/src/infra/simulator/mod.rs]: Good catch! I removed the `:?` since `eth::Gas` implements `Display` now :ok_hand:

File: crates/driver/src/domain/competition/mod.rs
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/mod.rs:39]: nit: should be merged into the one big `use` block we already have.

File: crates/driver/src/infra/simulator/mod.rs
Comment (MartinquaXD) [inline crates/driver/src/infra/simulator/mod.rs:155]: nit: let's be slightly more explicit here to reduce the chance that somebody has to explain to the solver teams what this error means.
```suggestion
    #[error("the simulated gas {0} exceeded the gas limit {1} provided in the solution")]
```

**Review by MartinquaXD — APPROVED** [review]
Before we merge this we probably have to let solvers know since the behavior of the same field in the response now is used for 2 different things between solving and quoting.
cc @harisang

**Review by squadgazzz — APPROVED** [review]

**Comment by harisang** [comment]
Just to clarify a few things:

1. This will be enforced to all solvers, or can we have a per solver flag? If it is enforced to all solvers, then I suspect there is no filtering if the "gas" field is missing from the response, right? In any case, if it is enforced to all solvers, then I would hold off for a few days so that I let all solvers know to either remove the field from their responses or report a crazy number so that it never fails (in case this is the behavior they aim for).
2. Does this have any effect on quotes? I would expect no, but wanted to clarify

File: crates/driver/src/domain/competition/mod.rs
Comment (sunce86) [inline crates/driver/src/domain/competition/mod.rs:327]: Why do we need to check this in our code? Isn't there a more elegant way to let simulator handle gas limit for us? tenderly, web3 and enso all should have a support for this.

This would require to add optional field `eth::Tx::gas` and forward the `settlement.gas.limit`

Comment (MartinquaXD) [reply crates/driver/src/domain/competition/mod.rs:327]: @Mateo-mro initially wanted to do it in the simulator as well but I recommended to do it in the domain logic. My reasoning was that overall if this is just intended to be a economic viability check it's more useful to know whether or not the solution works at all (e.g. to detect issues with the encoding) and only enforce the gas limit afterwards.

**Comment by Mateo-mro** [comment]
> Just to clarify a few things:
> 
> 1. This will be enforced to all solvers, or can we have a per solver flag? If it is enforced to all solvers, then I suspect there is no filtering if the "gas" field is missing from the response, right? In any case, if it is enforced to all solvers, then I would hold off for a few days so that I let all solvers know to either remove the field from their responses or report a crazy number so that it never fails (in case this is the behavior they aim for).
> 2. Does this have any effect on quotes? I would expect no, but wanted to clarify

1. It is enforced to all solvers
2. It shouldn't have any effect on quotes. This enforces it when simulating the settlement for `/settle`.
@harisang

File: crates/driver/src/domain/competition/mod.rs
Comment (Mateo-mro) [reply crates/driver/src/domain/competition/mod.rs:310]: Thanks @MartinquaXD for the answer! :pray: 

> This would require to add optional field eth::Tx::gas and forward the settlement.gas.limit

@sunce86  what do you exactly mean? :thinking:

**Comment by harisang** [comment]
Ok let me communicate this to solvers now and i guess we should be able to merge it in the next 2-days.

File: crates/driver/src/domain/competition/mod.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/mod.rs:310]: > @sunce86 what do you exactly mean? 

Just that the eth::Tx struct currently does not support providing gas limit field.

@MartinquaXD makes sense 👍

**Review by sunce86 — APPROVED** [review]

**Comment by Mateo-mro** [comment]
I received green light from @harisang . Merging it now.

**Pull request merged**
